core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -286,11 +286,31 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     tryOrElse(that.map(Right(_)), ZIO.succeedLeft)
 
   private final def tryOrElse[R1 <: R, E2, B](that: => ZIO[R1, E2, B], succ: A => ZIO[R1, E2, B]): ZIO[R1, E2, B] =
-    (self.tag: @switch) match {
-      case ZIO.Tags.Fail => that
+    new ZIO.Fold[R1, E, E2, A, B](
+      self,
+      cause => {
+        def firstDefect(c: Cause[_]): Throwable =
+          c match {
+            case c: Cause.Die => c.value
+            case c: Cause.Then[_] =>
+              val d = firstDefect(c.left)
+              if (d != null) d
+              else firstDefect(c.right)
+            case c: Cause.Both[_] =>
+              val d = firstDefect(c.left)
+              if (d != null) d
+              else firstDefect(c.right)
+            case _: Cause.Fail[_] | Cause.Interrupt =>
+              null
+          }
 
-      case _ => new ZIO.Fold[R1, E, E2, A, B](self, _ => that, succ)
-    }
+        val d = firstDefect(cause)
+
+        if (d != null) ZIO.die(d)
```

Comment:
I hate to ask for one more change, but I thought of a better way to do this:

Add `Cause#stripFailure: Cause[Nothing]`. 

Then you can do, `if (cause.died) ZIO.halt(cause.stripFailure)`.

The semantics would be: if there is no failure, it returns the same thing identically, but "proven" there is no failure (e.g. like the other method I mentioned); but if there is a failure, then it will strip it, but preserve all other information.

Then in the worst case, all we lose are one or more `E` values, but everything else is preserved.

Reply:
No worries! I was thinking in that direction as well.

Reply:
Changed to that formulation (except stripFailures had to be an `Option`) and added some tests to verify `died` and `stripFailures` are consistent.